internal/ssh/ssh.go

```
@@ -144,52 +142,36 @@ func directTCPIPHandler(sshManager SSHManager) func(srv *ssh.Server, conn *gossh
 			return
 		}
-		connInfo, err := sshManager.ControllerInfoFromModelUUID(ctx, modelTag.Id(), user)
-		if err != nil {
-			rejectConnectionAndLogError(ctx, newChan, "failed to get connection info", err)
-			return
-		}
-		client, err := dialControllerSSHServer(connInfo, d.DestPort)
+
+		client, err := sshManager.DialControllerSSHServer(ctx, modelTag.Id(), user)
 		if err != nil {
-			rejectConnectionAndLogError(ctx, newChan, fmt.Sprintf("failed to dial controller ssh: %v", err), err)
+			rejectConnectionAndLogError(ctx, newChan, "failed to dial controller", err)
 			return
 		}
 
-		dstChan, reqs, err := client.OpenChannel("direct-tcpip", gossh.Marshal(d))
+		controllerConn, err := client.Dial("tcp", fmt.Sprintf("%s:22", d.DestAddr))
 		if err != nil {
-			rejectConnectionAndLogError(ctx, newChan, "failed to open destination channel", err)
+			rejectConnectionAndLogError(ctx, newChan, "failed to create tunnel to controller", err)
 			return
 		}
-		// gossh.Request are requests sent outside of the normal stream of data (ex. pty-req for an interactive session).
-		// Since we only need the raw data to redirect, we can discard them.
-		go gossh.DiscardRequests(reqs)
 
-		srcDest, reqs, err := newChan.Accept()
+		clientConn, reqs, err := newChan.Accept()
 		if err != nil {
-			dstChan.Close()
+			rejectConnectionAndLogError(ctx, newChan, "failed to accept channel creation request", err)
 			return
 		}
-		// gossh.Request are requests sent outside of the normal stream of data (ex. pty-req for an interactive session).
-		// Since we only need the raw data to redirect, we can discard them.
 		go gossh.DiscardRequests(reqs)
 
 		go func() {
-			defer srcDest.Close()
-			defer dstChan.Close()
-			_, err := io.Copy(srcDest, dstChan)
-			if err != nil {
-				rejectConnectionAndLogError(ctx, newChan, "failed to copy data from src to dts", err)
-			}
+			defer clientConn.Close()
+			defer controllerConn.Close()
+			_, _ = io.Copy(clientConn, controllerConn)
```

why ignoring the error here?

---

Ah this is mb, I thought io.Copy returns an EOF error but I see the docstring says,
```
// A successful Copy returns err == nil, not err == EOF.
// Because Copy is defined to read from src until EOF, it does
// not treat an EOF from Read as an error to be reported.
```
So I'll revert this.

---

Although I think the way we handle the error is wrong, we can't reject the channel after accepting it.